File: app/src/main/java/com/fairshare/domain/model/sync/Snapshots.kt
package com.fairshare.domain.model.sync

import kotlinx.serialization.Serializable

/**
 * Wire-format snapshot of an [com.fairshare.domain.model.Event].
 *
 * Snapshots are decoupled from the in-app domain models so that the on-wire
 * representation stays stable even when the UI-facing model evolves. All
 * identifiers are [String] UUIDs (see DESIGN.md §2.1).
 */
@Serializable
data class EventSnapshot(
    val id: String,
    val name: String,
    val description: String? = null,
    val currency: String = "EUR",
    val createdAt: Long,
    /**
     * Archive flag. Travels through LWW like every other snapshot field:
     * the latest `EventUpsert` wins, so flipping it on any device
     * propagates via the standard sync pipeline. Defaulted to `false`
     * for wire-format backward compatibility (ignoreUnknownKeys is on,
     * older payloads without this field decode as archived = false).
     */
    val archived: Boolean = false,
)

/** Wire-format snapshot of a [com.fairshare.domain.model.Participant]. */
@Serializable
data class ParticipantSnapshot(
    val id: String,
    val eventId: String,
    val name: String,
)

/**
 * Wire-format snapshot of an [com.fairshare.domain.model.Expense].
 *
 * The whole tree (shares + items + assignments) is included because the chosen
 * CRDT granularity is per-Expense (see DESIGN.md §3.2). Concurrent edits on the
 * same expense resolve via LWW at this level.
 */
@Serializable
data class ExpenseSnapshot(
    val id: String,
    val eventId: String,
    val title: String,
    val amountCents: Long,
    val payerId: String,
    val date: Long,
    val shares: List<ExpenseShareSnapshot> = emptyList(),
    val items: List<ExpenseItemSnapshot> = emptyList(),
    /**
     * Marks the expense as a settlement (i.e. a real-world reimbursement
     * between two participants) rather than a shared cost. Settlements
     * are stored as plain expenses with payer = debtor and a single
     * share targeting the creditor, so the balance algorithm naturally
     * zeroes them out. The flag only changes the UI rendering. Defaulted
     * to `false` for wire-format backward compatibility.
     */
    val isSettlement: Boolean = false,
    /**
     * Optional category id tagging this expense. Either a
     * [com.fairshare.domain.model.DefaultCategories] stable id
     * (eg. `"default.food"`) or a UUID pointing to a custom
     * `CategorySnapshot` on the same event. Defaulted to `null` for
     * wire-format backward compatibility — old payloads without this
     * field decode as uncategorized.
     */
    val categoryId: String? = null,
)

/**
 * Wire-format snapshot of a [com.fairshare.domain.model.Category].
 *
 * Only custom (user-created) categories travel through the sync log;
 * default categories are hardcoded in [com.fairshare.domain.model.DefaultCategories]
 * and shared across devices by ID alone.
 */
@Serializable
data class CategorySnapshot(
    val id: String,
    val eventId: String,
    val name: String,
    val emoji: String,
    val color: Long,
)

@Serializable
data class ExpenseShareSnapshot(
    val id: String,
    val participantId: String,
    val amountCents: Long,
)

@Serializable
data class ExpenseItemSnapshot(
    val id: String,
    val label: String,
    val priceCents: Long,
    val quantity: Int = 1,
    /** Participant ids this item is assigned to. */
    val assignedTo: Set<String> = emptySet(),
)

File: webapp/src/core/sync/snapshots.ts
/**
 * Wire-format snapshots. Field names and default values must match
 * `com.fairshare.domain.model.sync.Snapshots.kt` byte-for-byte because
 * they cross the encrypted boundary and are read by Android peers.
 *
 * These are deliberately distinct from the in-app domain models (see
 * `core/domain/models.ts`) so the on-wire shape stays stable even as
 * the UI-facing model evolves.
 *
 * `kotlinx.serialization` with `encodeDefaults=true` emits every field
 * including nulls and defaults, in declaration order. The TS codec in
 * `codec.ts` reproduces that order on serialize and tolerates any
 * order on parse (`ignoreUnknownKeys=true` is symmetric).
 */

export interface EventSnapshot {
  id: string;
  name: string;
  description: string | null;
  currency: string;
  createdAt: number;
  archived: boolean;
}

export interface ParticipantSnapshot {
  id: string;
  eventId: string;
  name: string;
}

export interface ExpenseShareSnapshot {
  id: string;
  participantId: string;
  amountCents: number;
}

export interface ExpenseItemSnapshot {
  id: string;
  label: string;
  priceCents: number;
  quantity: number;
  /** Participant ids — kotlinx serialises `Set<String>` as a JSON array. */
  assignedTo: string[];
}

export interface ExpenseSnapshot {
  id: string;
  eventId: string;
  title: string;
  amountCents: number;
  payerId: string;
  date: number;
  shares: ExpenseShareSnapshot[];
  items: ExpenseItemSnapshot[];
  isSettlement: boolean;
  categoryId: string | null;
}

export interface CategorySnapshot {
  id: string;
  eventId: string;
  name: string;
  emoji: string;
  color: number; // ARGB
}
